cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,473 @@
+
+##### Pod Startup
+* Similarly to holding Pod startup until after ensuring the ipcache is up to date, once mTLS is enabled in the cluster, we'll need to hold up Pod startup while certificates are provisioned. Still TBD is if this will be for _every_ Pod or only for Policy-relevant ones.
+
+##### CiliumNetworkPolicy creation
+Parse CNPs for the new authentication field. When this is present, as part of policy plumbing, set a new field in the policymap entries that indicates that authentication is required for connectivity to this peer.
+This process will be skipped if the `mesh-auth-mode` config directive is not set.
+
+##### Triggered by datapath drop
+* When the datapath generates a datapath drop message with the reason that "authentication is required" for this peer, look up the corresponding node and establish an mTLS connection to the peer node. Connects to the port outlined in the Startup section above.
+  * Use the Cilium Identity to look up TLS details.
+  * Special case: peer node may be the same node as the current node. In this case, the Node can connect to its own serving port, and perform the handshake, just like a remote node.
+  * Validate certificate of the peer. Note that since we are also planning to handle re-key events by pre-emptively re-authenticating, this should not occur very often.
+  Intial implementation efforts show that when using SPIRE, the SPIRE server will pre-emptively re-key for you before the expiration time has elapsed, and send a new keypair via the DelegatedIdentity API - we are planning on triggering reauth when this occurs.
+    * Q: How to make invalid certificate errors super obvious for debugging?
+      * Logs
+      * Metrics
+
+##### Authenticated session established with peer
+
+* Update datapath to inform it that this peer is authenticated (see Data Plane for details)
+* The destination node is responsible for checking that a Pod matching the destination Identity is running on itself, and only allowing the authentication to succeed if it is.
+
+##### Endpoint deletion
+
+* Node with endpoint must clear out Auth table to remove any state that implies authentication occurred for connections with this endpoint
+* Do all other nodes need to clear out their corresponding state also?
+  * Thinking of IP recycling case, e.g. in ENI mode an IP could move from one node to another.
+  * Trigger on CiliumEndpoint deletion even for CiliumEndpoints on remote nodes, delete authentication state corresponding to that peer. Including userspace authentication table entries, datapath CT/map entries.
+
+##### Timer-based
+* Do we need periodic garbage collection, e.g. of authenticated sessions? Because the auth table has an expiry time, we have built-in garbage collection - the userspace will be responsible for pruning expired auth table entries if the identity has been deleted and the certificate validity period has passed.
```

Comment:
nit: The question is answered by saying that garbage collection is "built-in", but I don't quite understand where that assumption comes from. The current section is "Local Agent Events -> "Timer-based". When does this "built-in" garbage collection run exactly?

(I assumed there would be a list sorted by expiry time, then a GC thread that just looks at the list, sleeps until the next expiry, wakes up, performs GC, then sleeps again until the next expiry time, maybe with a minimum sleep period or something)

Reply:
If this expiry is "built-in" by some sort of event, such as waiting on an external system to trigger an exiry timer, then maybe we should rework this "Timer-based" section to say "Upon auth expiry event". The important part there is that we are then relying on the authentication module (SPIFFE libs?) to perform that callback.

Reply:
I've updated, but the idea here is that because the auth table entry includes an expiry time (after which the data path will drop packets that match it), garbage collection is not _as_ pressing a concern. Currently, we're planning on having userspace remove entries once the identity goes away.

Reply:
Makes sense yep, I'm just asking "How exactly with userspace remove entries?" - what event causes userspace to evaluate which entries to remove and what do we need to do to track those. I'm sure we'll figure this out over time.